repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala

```diff
@@ -113,27 +107,35 @@ class HttpServletRequestWrapper(originalRequest: HttpServletRequest, inputStream
     headerMap = headerMap - headerName
   }
 
-  def getPreferredHeader(headerName: String, getFun: String => List[String]): String = {
+  private def getPreferredHeader(headerName: String, getFun: String => List[String]): List[String] = {
+    case class HeaderValue(value: String, quality: Double)
+
+    def parseValue(headerValue: String): String = headerValue.split(";").head
+
     def parseQuality(headerValue: String): Double = {
       try {
         val headerParameters: Array[String] = headerValue.split(";").tail
         val qualityParameters: Option[String] = headerParameters.find(param => "q".equalsIgnoreCase(param.split("=").head.trim))
         qualityParameters.map(_.split("=", 2)(1).toDouble).getOrElse(1.0)
-      }
-      catch {
+      } catch {
         case e: NumberFormatException => throw new QualityFormatException("Quality was an unparseable value", e)
       }
     }
 
     getFun(headerName) match {
       case Nil => null
-      case nonEmptyList => nonEmptyList.maxBy(parseQuality).split(";").head
+      case nonEmptyList =>
+        nonEmptyList.map(headerValue => HeaderValue(parseValue(headerValue), parseQuality(headerValue))) // split the value and parameters, parse the quality
+          .groupBy(_.quality) // group by quality
+          .maxBy(_._1) // find the highest quality group
```

Comment:
Probably, but I'm not sure it would be any cleaner. If you've got something in mind we can change this.